### PR TITLE
VACMS-7011: Move help text below labels.

### LINKED ISSUE #7011: Move text descriptions for all long text fields above the field
## User Story or Problem Statement

As an editor of a long text field, i want guidance _before_ i write my content, not after it, so that i don't waste my time, and that i create content that fits the content strategy based on the needs of Veterans. 

### Context

Drupal puts help text for text areas after the field, by default. 

We have 54 textarea fields. Most of them have help text. (aka field "descriptions"). 25 out of 54 also have textcounters, which shows below the field, further displacing the help text. 

<details><summary>Example</summary>

Edit_Vet_Center_Kona_Vet_Center___VA_gov_CMS

</details>

## Acceptance Criteria
- [ ] We have a POC set that moves the help text for textareas above the field for all textarea fields
  - [ ] design review (spacing/hierarchy, unintended impacts, etc)
  - [ ] change management review 
- [ ] Textcounters remain below the field

## Implementation steps
Field widget twig theming? 

## Design principles

Veteran-centered
- [ ] `Single source of truth`: Increase reliability and consistency of content on VA.gov by providing a single source of truth.
- [ ] `Accessible, plain language`: Provide guardrails and guidelines to ensure content quality.
- [ ] `Purposely structured content`: Ensure Content API can deliver content whose meaning matches its structure.
- [ ] `Content lifecycle governance`: Produce tools, processes and policies to maintain content quality throughout its lifecycle.

Editor-centered
- [x] `Purpose-driven`: Create an opportunity to involve the editor community in VA’s mission and content strategy goals.
- [ ] `Efficient`: Remove distractions and create clear, straightforward paths to get the job done.
- [ ] `Approachable`: Offer friendly guidance over authoritative instruction.
- [x] `Consistent`: Reduce user’s mental load by allowing them to fall back on pattern recognition to complete tasks.
- [x] `Empowering`: Provide clear information to help editors make decisions about their work.

### CMS Team

Please leave only the team that will do this work selected. If you're not sure, it's fine to leave both selected.

- [ ] `Platform CMS Team`
- [x] `Sitewide CMS Team`

### PR DESCRIPTION
**Pull request by ethanteague**
## Description
closes #7011

## Testing done
Visual

## Screenshots
image
image

## QA steps
 - [ ] As an administrator, go to `node/add/page` & `node/add/landing_page` and visually verify help text appears below labels for all text and wysiwyg form fields.

### CODE REVIEW

File: docroot/modules/custom/va_gov_backend/js/script.js [hunk 1]
```diff
@@ -40,7 +40,7 @@
         });
 
         var adminRoles = ["content_admin", "administrator"];
-        var targetTypes = ["health_care_local_facility", "vet_center"];
```
Comment (ethanteague) [inline docroot/modules/custom/va_gov_backend/js/script.js:43]: Pulled this out without compiling in a previous pr. Removal expected.

**Comment by kevwalsh** [comment]
**Overall approach and QA challenges**

I ❤️ that the scope of this is growing. But it also means we need closer QA and design review. 

Originally, this was just going to be fields marked "Text (formatted, long)" and Text "(plain, long)". Now it includes all "List (text)", "Entity reference", and a bunch of others. 

I just did some comparison of prod and this PR environment, just of the fields types present on Content types (not on paragraphs, which may have some field types that paragraphs do not) and this is what i found to be changing. 


CMS_DS_Component___Pattern_inventory__Field_types_-_Airtable

It just occurred to me that this may not be by Field type, but may be a bit more nuanced, by the Field widget or some other attribute of each field instance. Like for entity references, there a dozen different field formatters: 

> Select list, Entity browser, Inline entity form - Complex, Autocomplete, Entity Browser - Table, Media library, Inline entity form - Complex - Table View Mode, Check boxes/radio buttons.

I didn't check all of these. 

It's also unclear how many core field-type-things, not known to our Drupal Spec Tool, like URL alias, may be affected by this. @ethanteague you might have a sense of how far-ranging this could be. 




**Design review**

Getting a rough sample of before/after would probably be helpful for design review. 

My sense is that the help text needs a bit more `padding-bottom`, but there may be other things to watch for. 


**Change management** 

I'd like to get @xiongjaneg's  eyes on this to help determine the impact this might have on users, not to mention that this will immediatlely "date" some existing training and KB materials. Not as much as Claro will. 

**Some specific issues i've seen, not necessarily deal-breakers**


<details><summary>Phone number fields</summary>

We could solve this by dropping the extension below, but there may be other ways.

Cursor_and_Create_VAMC_Facility_Health_Service___VA_gov_CMS


</details>

**Review by swirtSJW — COMMENTED** [review]
Code looks fine but the QA steps seem a little light for this sweeping of a change.

**Comment by ethanteague** [comment]
> Code looks fine but the QA steps seem a little light for this sweeping of a change.  

Agreed buddy. I think we need a swarm session.

**Comment by kevwalsh** [comment]
A few followups from Thursday 9:15am

- [ ] Padding
- [ ] Check all the widget types from [this airtable(https://airtable.com/appKn8xeWg0SkIphM/tblyDT3IsWT9g4V3h/viwOB6i525yBSDZFF?blocks=hide), like IEF for Entity reference
- [ ] Check Integer and office hours (may need help text to test) 
- [ ] Phone number extension wrap
- [ ] Date field (see https://prod.cms.va.gov/node/add/va_form) 
- [ ] Improve comments in the code for how the patterns work and relate to field or widget types

File: docroot/modules/custom/va_gov_backend/js/script.es6.js [hunk 2]
```diff
@@ -284,4 +284,85 @@
...
+      const textareaDescriptions = context.querySelectorAll(".description");
+      if (textareaDescriptions.length > 0) {
+        textareaDescriptions.forEach((item) => {
+          // Moves help text from below field / widget to below label.
+          // This approach relies on html structure patterns to broadly
+          // target fields and widgets. It is field name / type agnostic.
+          // The patterns below have comments that indicate their general scope.
+
+          // Date fields.
+          if (
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/js/script.es6.js:300]: Nested nested ifs are a bit long.  Can we compartmentalize each one to an appropriately named function that does the checking and alters if true
example: 
alterHelpTextLocationOnDateField
alterHelpTextLocationOnTextField
alterHelpTextLocationOnLongTextFullHtmlField
alterHelpTextLocationOnLongTextPlainField ...

It makies the code easier to follow... especially on ES6 which ends up with no comments.  Makes it less east to break, and more testable.

**Review by swirtSJW — CHANGES_REQUESTED** [review]
Overall good.  One sustainability request.

File: docroot/modules/custom/va_gov_backend/js/script.es6.js (hunk 2 above)
Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/js/script.es6.js:300]: Good thinking, @swirtSJW - will do :+1:

Comment (ethanteague) [reply docroot/modules/custom/va_gov_backend/js/script.es6.js:300]: Think I'll move this out into it's own library, as well.

**Review by msbtterswrth — CHANGES_REQUESTED** [review]
@ethanteague can we make one TINY change? I think the spacing is too much between the label and the description now, let's adjust it to `margin: 0 0 10px 0`.

image

**Comment by ethanteague** [comment]
Updated, @msbtterswrth @swirtSJW :+1:

**Review by msbtterswrth — APPROVED** [review]
Looks perf! Great job!

File: docroot/modules/custom/va_gov_backend/js/help_text_treatments.es6.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      // Date fields.
+      alterHelpTextLocationOnDateField(textareaDescriptions);
+      // Text fields.
+      alterHelpTextLocationOnTextField(textareaDescriptions);
+      // Long text fields (No Wysiwyg).
+      alterHelpTextLocationOnLongTextPlainField(textareaDescriptions);
+      // Long text field full_html (Wysiwyg).
+      alterHelpTextLocationOnLongTextFullHtmlField(textareaDescriptions);
+      // Paragraphs experimental widget.
+      alterHelpTextLocationOnParagraphsExperimentalWidget(textareaDescriptions);
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/js/help_text_treatments.es6.js:120]:
```suggestion
      // Get help texts.
      const textareaDescriptions = context.querySelectorAll(".description");
      alterHelpTextLocationOnDateField(textareaDescriptions);
      alterHelpTextLocationOnTextField(textareaDescriptions);
      alterHelpTextLocationOnLongTextPlainField(textareaDescriptions);
      alterHelpTextLocationOnLongTextFullHtmlField(textareaDescriptions);
      alterHelpTextLocationOnParagraphsExperimentalWidget(textareaDescriptions);
```
With the good naming of the methods that you did, these comments are unnecessary.

File: docroot/modules/custom/va_gov_backend/js/help_text_treatments.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+      alterHelpTextLocationOnDateField(textareaDescriptions);
+
+      alterHelpTextLocationOnTextField(textareaDescriptions);
+
+      alterHelpTextLocationOnLongTextPlainField(textareaDescriptions);
+
+      alterHelpTextLocationOnLongTextFullHtmlField(textareaDescriptions);
+
+      alterHelpTextLocationOnParagraphsExperimentalWidget(textareaDescriptions);
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_backend/js/help_text_treatments.js:78]:
```suggestion
      var textareaDescriptions = context.querySelectorAll(".description");
      alterHelpTextLocationOnDateField(textareaDescriptions);
      alterHelpTextLocationOnTextField(textareaDescriptions);
      alterHelpTextLocationOnLongTextPlainField(textareaDescriptions);
      alterHelpTextLocationOnLongTextFullHtmlField(textareaDescriptions);
      alterHelpTextLocationOnParagraphsExperimentalWidget(textareaDescriptions);
```

**Review by swirtSJW — COMMENTED** [review]
super minor fixes i am going to apply.

**Review by swirtSJW — APPROVED** [review]
Nice work.

**Review by swirtSJW — APPROVED** [review]

**Comment by swirtSJW** [comment]
image

**Pull request merged**